Add Empress class and add support for OrdinationResults in q2-empress

ElDeveloper opened the pull request:
This PR moves over the code from _plot.py into a new class `Empress`. In addition this allows support for visualizing ordinations and phylogenies simultaneously. Here's an animated gif of this working with brush selection (related to https://github.com/biocore/emperor/pull/766).

![empress-pr](https://user-images.githubusercontent.com/375307/84715877-a3cd9a00-af26-11ea-8e6d-0bcd74f34fa2.gif)

The example above shows how selecting a group of samples of different colors shows the unique branches per color (similar to UniFrac comparing groups). When a single sample or group is selected, all the features for that group are highlighted in Empress.

`qiime empress plot` now accepts an optional ordination object.
The majority of the code for the Empress class was inspired by the Emperor object.

fedarko commented on empress/support_files/js/empress.js:
```diff
@@ -546,10 +546,58 @@ define([
+            obs = Array.from(observationsPerGroup[group]);
+
+            // convert hex string to rgb array
+            var rgb = [
+                parseInt(group.slice(0, 2), 16) / 255,
+                parseInt(group.slice(2, 4), 16) / 255,
+                parseInt(group.slice(4), 16) / 255,
+            ];
```
Suggest delegating this to chroma.js rather than doing it manually -- `chroma(color).gl().slice(0, 3);` is how the `Colorer` module in Empress converts hex to RGB.

Of course the code is still correct as is, so this doesn't matter too much...

ElDeveloper replied:
Good call!

ElDeveloper commented on empress/core.py:
```diff
@@ -0,0 +1,335 @@
+        with open(SELECTION_CALLBACK_PATH) as f:
+            selection_callback = f.read()
+        emperor_require_logic = emperor_require_logic.replace(
+            '/*__select_callback__*/', selection_callback)
```
I guess technically there's a slim chance this could happen. In general, I think we'll want to make it possible for the `Emperor` class to be able to render the custom code automatically and not need to do these string manipulations directly. This should be part of improving that API.

ElDeveloper commented:
Thanks @fedarko, all changes should be addressed.

On (Jun-16-20|14:58), Marcus Fedarko wrote:
>@fedarko requested changes on this pull request.
>
>A couple minor suggestions, but this looks really nice :)
>
>> @@ -42,13 +44,14 @@
>     },
>     input_descriptions={
>         'tree': 'The phylogenetic tree to visualize.',
>-        'feature_table': (
>-            'A table containing the abundances of features within samples. '
>-            'This information allows us to decorate the phylogeny by '
>-            "sample metadata. It's expected that all features in the table "
>-            'are also present as tips in the tree, and that all samples in '
>-            'the table are also present in the sample metadata file.'
>-        )
>+        'feature_table': 'A table containing the abundances of features within'
>+                         ' samples. This information allows us to decorate the'
>+                         ' phylogeny by sample metadata. It\'s expected that '
>+                         ' all features in the table are also present as '
>+                         'tips in the tree, and that all samples in the table'
>+                         ' are also present in the sample metadata file.',
>
>Minor, but there are extra spaces included in some of the lines here. It would be nice to make this consistent, with trailing spaces at the end of every string except the last one.
>
>> @@ -1,8 +1,9 @@
> # ----------------------------------------------------------------------------
>-# Copyright (c) 2018--, empress development team.
>+# Copyright (c) 2016-2020, QIIME 2 development team.
>
>I don't think the 3-4 people working on Empress right now count as the "QIIME 2 development team"... I think keeping this as "empress development team" would make more sense
>
>> +        Notes
>+        -----
>+        Once you generate the plot you will need to copy the support files by
>+        calling the ``copy_support_files`` method.
>+
>+        See Also
>+        --------
>+        empress.core.Empress.copy_support_files
>+        """
>+        main_template = self._get_template()
>+
>+        # _process_data does a lot of munging to the coordinates data and
>+        # _to_dict puts the data into a dictionary-like object for consumption
>+        data = self._to_dict()
>+
>+        # need to do something about low and high
>
>Not sure what this comment means?
>
>> +        filter_missing_features: bool, optional
>+            If True, filters features from the table that aren't present as
>+            tips in the tree. If False, raises a DataMatchingError if any such
>+            features exist. (Note that in either case, features in the tree but
>+            not in the table are preserved.)
>+        resource_path: str, optional
>+            Load the resources from a user-specified remote location. If set to
>+            None resources are loaded from the current directory.
>+
>+
>+        Attributes
>+        ----------
>+        tree:
>+            Phylogenetic tree.
>+        table:
>+            Contingency matrix for teh phylogeny.
>
>```suggestion
>            Contingency matrix for the phylogeny.
>```
>
>> +            not in the table are preserved.)
>+        resource_path: str, optional
>+            Load the resources from a user-specified remote location. If set to
>+            None resources are loaded from the current directory.
>+
>+
>+        Attributes
>+        ----------
>+        tree:
>+            Phylogenetic tree.
>+        table:
>+            Contingency matrix for teh phylogeny.
>+        samples:
>+            Sample metadata.
>+        features:
>+            Feature metadata
>
>```suggestion
>            Feature metadata.
>```
>
>> +            DataFrame object with the metadata associated to the features in
>+            the ``table`` and ``tree`` objects, should have an index set and it
>+            should match the feature identifiers.
>
>```suggestion
>            DataFrame object with the metadata associated to the names of
>            tips and/or internal nodes in the  ``tree`` object, should have an
>            index set and it should match at least one of these nodes' names.
>```
>
>(modified this to better describe the way feature metadata is implemented, feel free to alter as you see fit)
>
>> +            DataFrame object with the metadata associated to the features in
>+            the ``table`` and ``tree`` objects, should have an index set and it
>+            should match the feature identifiers.
>+        ordination: skbio.OrdinationResults, optional
>+            Object containing the computed values for an ordination method in
>+            scikit-bio. Currently supports skbio.stats.ordination.PCoA and
>+            skbio.stats.ordination.RDA results.
>+        ignore_missing_samples: bool, optional
>+            If True, pads missing samples (i.e. samples in the table but not
>+            the metadata) with placeholder metadata. If False, raises a
>+            DataMatchingError if any such samples exist. (Note that in either
>+            case, samples in the metadata but not in the table are filtered
>+            out; and if no samples are shared between the table and metadata, a
>+            DataMatchingError is raised regardless.) This is analogous to the
>+            ignore_missing_samples flag in Emperor.
>+        filter_missing_features: bool, optional
>
>```suggestion
>        filter_missing_features: bool, optional (default False)
>```
>
>> +            The phylogenetic tree to visualize.
>+        table: pd.DataFrame:
>+            The matrix to visualize paired with the phylogenetic tree.
>+        sample_metadata: pd.DataFrame
>+            DataFrame object with the metadata associated to the samples in the
>+            ``ordination`` object, should have an index set and it should match
>+            the identifiers in the ``ordination`` object.
>+        feature_metadata: pd.DataFrame, optional
>+            DataFrame object with the metadata associated to the features in
>+            the ``table`` and ``tree`` objects, should have an index set and it
>+            should match the feature identifiers.
>+        ordination: skbio.OrdinationResults, optional
>+            Object containing the computed values for an ordination method in
>+            scikit-bio. Currently supports skbio.stats.ordination.PCoA and
>+            skbio.stats.ordination.RDA results.
>+        ignore_missing_samples: bool, optional
>
>```suggestion
>        ignore_missing_samples: bool, optional (default False)
>```
>
>> +        Once you generate the plot you will need to copy the support files by
>+        calling the ``copy_support_files`` method.
>
>```suggestion
>        Once you generate the plot (and write it to a HTML file in a given directory)
>        you will need to copy the support files (the JS/CSS/etc. code needed to
>        view the visualization) to the same directory by calling the
>        ``copy_support_files`` method.
>```
>
>> +        # line 14 is where the CSS includes start, but it is surrounded by
>+        # unnecessary tags so we strip those out
>+        style = '\n'.join([line.strip().replace("'", '').replace(',', '')
>+                           for line in html[14:20]])
>+
>+        # main divs for emperor
>+        emperor_div = '\n'.join(html[39:44])
>+
>+        # main js script for emperor
>+        emperor_require_logic = '\n'.join(html[45:-3])
>+
>+        # once everything is loaded replace the callback tag for custom JS
>+        with open(SELECTION_CALLBACK_PATH) as f:
>+            selection_callback = f.read()
>+        emperor_require_logic = emperor_require_logic.replace(
>+            '/*__select_callback__*/', selection_callback)
>
>Do you think there is a chance of `/*__select_callback__*/` (or other "reserved" things to be replaced in the HTML) accidentally being used in multiple places? (Like, would it be possible for a user to name one of their samples that?)
>
>If there is a chance (even a slim one), it may be worth opening another issue for later on to screen datasets for these kinds of things and raise an error accordingly. (I'm realizing as I write this now that I guess this could happen with the current jinja templates in Empress, if the user has something in their data named `{{ tree_data }}` or whatever.)
>
>> +            var rgb = [
>+                parseInt(group.slice(0, 2), 16) / 255,
>+                parseInt(group.slice(2, 4), 16) / 255,
>+                parseInt(group.slice(4), 16) / 255,
>+            ];
>
>Suggest delegating this to chroma.js rather than doing it manually -- `chroma(color).gl().slice(0, 3);` is how the `Colorer` module in Empress converts hex to RGB.
>
>Of course the code is still correct as is, so this doesn't matter too much...
>
>-- 
>You are receiving this because you authored the thread.
>Reply to this email directly or view it on GitHub:
>https://urldefense.com/v3/__https://github.com/biocore/empress/pull/174*pullrequestreview-431674162__;Iw!!Mih3wA!XKWTgkOanhxlTCIZ0bLGJEBAu5X_uj86COJ7uzhNFKKQFL8wpWtO2tDv2zPGQlU$

wasade commented:
this looks so cool....

kwcantrell commented:
Looks Good! 

I think this is outside of the scope of this PR but if the user had colored the empress tree using the sample coloring feature and then selects nodes in the emperor plot, the color by check box remains checked.
![emperor](https://user-images.githubusercontent.com/28127694/84838315-ee1a4e00-afee-11ea-9000-94fdd4d94bd7.png)
 
It would be nice to either reset the sample coloring panel or show the update button like you get when you change a coloring category as shown below. Otherwise if the user wants to recolor the tree again they would have to uncheck and then recheck the color by check box.
![emperor2](https://user-images.githubusercontent.com/28127694/84838399-30438f80-afef-11ea-9e20-f1d976f22464.png)

fedarko commented on empress/core.py:
```diff
@@ -0,0 +1,335 @@
+        # _to_dict puts the data into a dictionary-like object for consumption
+        data = self._to_dict()
+
+        # need to do something about low and high
```
@ElDeveloper ... I still don't know what this comment means :P

fedarko commented:
Thanks @ElDeveloper! Agreed with Daniel and Kalen that this is awesome.

I tried this out locally and have played around with it -- I have a few remaining comments (if we could address them today that'd be awesome, if not then I guess we can get to them later).

1. I had to update my Emperor install to get this to work (which makes sense). However, it might be worth adding a check at some point that looks at the Emperor version and verifies that it's new enough to work with this? If an outdated version is in use (and we should keep in mind that 99.99% of people using these tools will have an "outdated" version of Emperor since the newest changes haven't been released yet), then we should raise an error or something explaining this. (We could even impose a minimum Emperor version in setup.py?)

2. Using a biplot as `--i-pcoa` works, but what happens is that the created version of Emperor doesn't seem to have a limit on the number of arrows shown (see below). If possible, changing the Emperor defaults on "number of arrows shown" to match the q2-Emperor defaults would be nice. If this wouldn't be ok, then we should probably expose a `--p-number-of-arrows` or something for Empress.

![plot](https://user-images.githubusercontent.com/4177727/84838481-65e87880-afef-11ea-8383-cbc4f916270b.png)

3. Based on [this thread](https://github.com/biocore/emperor/pull/766#discussion_r439093981), I tried selecting samples from a parallel plot to view in Empress -- but it didn't seem like the samples were passed to Empress this time. Is this expected?

4. When I view the Emperor plot in 2-D and right-click drag to translate the display around, I can't seem to move the display up or down -- I can only move it left or right. Weirdly, when I add back Axis 3 (making the plot 3-D), the problem is still here -- but it sort of goes away when you tilt the plot along Axis 1. There seems to be some sort of "resistance" to moving the plot around that is somehow dependent upon the direction Axis 3 is pointing. My guess is that something is messing with the camera in Emperor. (This is a bizarre, unimportant problem, so it may be worth shelving and addressing later.)

5. (Not within the scope of this PR, but just for reference) I should also point out that the #152 bug is present here as well (which makes sense since the JS code also uses `_projectObservations()`), so once that is fixed this should highlight unique sections of the tree properly:

![image](https://user-images.githubusercontent.com/4177727/84837588-5cf6a780-afed-11ea-8bd8-6701e5fe72a7.png)

kwcantrell commented:
@ElDeveloper so you think we should still propagate the color up the tree if an internal node has children that do not belong to any samples (i.e. we will still get the current behavior for tips that are not in the feature table) and don't propagate the color when an internal node has children that belong to multiple samples? I think this option is better. However, we should distinguish uncolored tip that belong to multiple samples vs tips that aren't in the feature table. So, I think we should add all tips that belong to multiple categories to a "non-unique" category and assign them a color.

kwcantrell commented on empress/support_files/js/selection-callback.js:
```diff
@@ -0,0 +1,33 @@
+    }
+    colorGroups[color].push(samples[i].name);
+}
+empress.colorSampleGroups(colorGroups);
+
+// 3 seconds before resetting
+setTimeout(function () {
+    empress.resetTree();
+    empress.drawTree();
+
+    samples.forEach(function (sample) {
+        sample.material.emissive.set(0x000000);
+    });
```
I think it might be better to trigger the reset operation when the updates either the emperor plot or empress tree. That way the user explore the empress tree after the selected nodes in the emperor plot.

ElDeveloper replied:
While I agree that this would be best, we can't really do that at the moment without linking additional empress and emperor callbacks (or am I missing something?).

kwcantrell replied:
Okay I'll open up an issue.

ElDeveloper commented:
@fedarko:

1) I've added installation instructions to the README, which is I think the best we can do for the time being.

2) I've added the argument as is from q2-emperor.

3) Turns out this is a bug that I can't quite fix with empress. This needs to be fixed in Emperor: https://github.com/biocore/emperor/issues/771

4) Thanks for opening an issue in emperor https://github.com/biocore/emperor/issues/770

5) Should be addressed by Kalen's PR.

ElDeveloper commented:
@kwcantrell, and @fedarko thanks so much you two! 🚀
